### PR TITLE
FIRE-30873: Add Poser

### PR DESCRIPTION
**Pull request by AngeldarkRaymaker**
Hi there!
Get snacks! Long ride.

There's two bodies of work: the Poser, originally from BD, but I got it from Alchemy, because I thought that might make it easier on me. This adds the 'Poser' and the 'Pose Creator', the later allowing import into SL.
Secondly, I did a little work on llSliders: they're a bit tedious to use as-is for Poser, they need easy revert and the ability for fine-grained control.
This also changes behaviour of what ctrl-click does on a slider: it used to set it back to the first thing it was. Now it resets it back to what it was since you last moved it.

I've put no <FS> tags in the black-dragon code, as it is not LL's work.

Build-wise it's not clear to me if the new XML's (for floaters, etc) need to be added somewhere to be included in the distribution.
Let me know.

There's one hack in there, see comment.

I've also added a tga file: these appear deprecated elsewhere, I could do something else.

All images are copies from others in this repo.

Presently, the Poser is self-only.

I'd like to continue work to have the LSL Bridge request permission to animate other avatars using the normal LLDialog. Not that the other avatar could see it (unless... ha ha!), but, that would make this tool bananas compared to other offerings.
I think this would fit with current LSL permissions: requiring opt-in from the avatar to be animated.

## Firestorm Pull Request Checklist
Thank you for contributing to the Phoenix Firestorm Project.
We will endeavour to review you changes and accept/reject/request changes as soon as possible. 
Please read and follow the [Firestorm Pull Request Guidelines](https://github.com/firestormviewer/phoenix-firestorm/blob/master/FS_PR_GUIDELINES.md) to reduce the likelihood that we need to ask for "Bureaucratic" changes to make the code comply with our workflows.

### CODE REVIEW

File: indra/llui/llslider.cpp [hunk 1]
```diff
@@ -279,12 +281,31 @@ BOOL LLSlider::handleKeyHere(KEY key, MASK mask)
 
 BOOL LLSlider::handleScrollWheel(S32 x, S32 y, S32 clicks)
 {
-    if ( mOrientation == VERTICAL )
+    // <FS> [FIRE-30873]:  BD - UI Improvements: wheel works on both axes with ctrl, ctrl+shift+wheel moves at 10% sensitivity
+    SHORT shiftKeyState = GetAsyncKeyState(VK_SHIFT);
+    SHORT ctrlKeyState = GetAsyncKeyState(VK_CONTROL);
```
Comment (AngeldarkRaymaker) [inline indra/llui/llslider.cpp:286]: These are hacky.
I looked at injecting MASK, and tried, but boy, I could have gone all the way back to base, allowing mask to be offered everywhere... which isn't a bad thing.

What I worry about is if this is platform dependent. I think Mac users won't like CTRL.

**Comment by beqjanus** [comment]
Thank you very for this, and for the diligence in the merge such as the markers for future enhancements.

I've built this locally, and the UI is not appearing as I had expected.
Is this what you see?
Notably, we have a large grey square that pushes the avatar name off the side.
We also have a similar spacing issues on the bone list. 
db05402947f14f7ef74ad5d828389b15

File: indra/newview/bdanimator.cpp [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+}
+
+BDAnimator::~BDAnimator()
+{
+}
+
+void BDAnimator::update()
+{
+    //BD - Don't do anything if the animator is not activated.
+    if (!mPlaying) return;
```
Comment (Ansariel) [inline indra/newview/bdanimator.cpp:52]: I don't like the if-clause and the true-statement being in one line

File: indra/newview/bdanimator.cpp [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+        avatar->mAnimPlayTimer.stop();
+    }
+    mPlaying = false;
+}
+
+//BD - We allow loading rotations, positions and scales seperately
+//     by giving the load an integer which determines what to load.
+//     1 is default and loads rotations only, 2 = positions only,
+//     4 = scales only, thus 3 = rotations and positions and so on.
+BOOL BDAnimator::loadPose(const LLSD& name, S32 load_type)
```
Comment (Ansariel) [inline indra/newview/bdanimator.cpp:244]: Should be a true bool and not BOOL

File: indra/newview/bdanimator.cpp [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+                    motion->setInterpolationType(type);
+                    motion->setInterpolationTime(time);
+                    motion->startInterpolationTimer();
+                }
+
+                LLJoint* joint = mTargetAvatar->getJoint(name);
+                if (joint)
+                {
+                    //BD - Don't try to add/remove joint states for anything but our default bones.
+                    if (motion && joint->getJointNum() < 134)
```
Comment (Ansariel) [inline indra/newview/bdanimator.cpp:304]: What is 134? Can we turn this into a constexpr with a meaningful name?

File: indra/newview/bdanimator.cpp [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+            }
+        }
+        else
+        {
+            LLJoint* joint = mTargetAvatar->getJoint(pose["bone"].asString());
+            if (joint)
+            {
+                BDPosingMotion* motion = (BDPosingMotion*)mTargetAvatar->findMotion(ANIM_BD_POSING_MOTION);
+                //BD - Don't try to add/remove joint states for anything but our default bones.
+                if (motion && joint->getJointNum() < 134)
```
Comment (Ansariel) [inline indra/newview/bdanimator.cpp:373]: The magic 134 again...

File: indra/newview/bdfloaterposecreator.cpp [hunk 6]
```diff
@@ -0,0 +1,3067 @@
...
+}
+
+////////////////////////////////
+//BD - Misc Functions
+////////////////////////////////
+LLKeyframeMotion* BDFloaterPoseCreator::onReadyTempMotion(std::string filename, bool eternal)
+{
+    std::string outfilename = gDirUtilp->getExpandedFilename(LL_PATH_USER_SETTINGS, "animations", filename);
+    S32 file_size;
+    BOOL success = FALSE;
```
Comment (Ansariel) [inline indra/newview/bdfloaterposecreator.cpp:2819]: Use true bool instead of BOOL

File: indra/newview/bdfloaterposecreator.cpp [hunk 7]
```diff
@@ -0,0 +1,3067 @@
...
+    std::vector<LLPointer<LLJointState>> joint_states;
+
+    if (!mTempMotionList)
+        return;
+
+    if (!mTempMotionList->mJointMotionArray.empty())
+        mTempMotionList->mJointMotionArray.clear();
+    mTempMotionList->mJointMotionArray.reserve(134);
+    joint_states.clear();
+    joint_states.reserve(134);
```
Comment (Ansariel) [inline indra/newview/bdfloaterposecreator.cpp:2901]: The magic 134 is back...

File: indra/newview/bdfloaterposecreator.cpp [hunk 8]
```diff
@@ -0,0 +1,3067 @@
...
+    std::vector<std::string> joint_names, cv_names, attach_names;
+    gAgentAvatarp->getSortedJointNames(0, joint_names);
+    gAgentAvatarp->getSortedJointNames(1, cv_names);
+    gAgentAvatarp->getSortedJointNames(2, attach_names);
+
+    mDefaultRotations.clear();
+    mDefaultScales.clear();
+    mDefaultPositions.clear();
+
+    for (auto name : joint_names)
```
Comment (Ansariel) [inline indra/newview/bdfloaterposecreator.cpp:3015]: Why not `for (const auto& name : joint_names)`? Pretty sure I there are other cases in this file as well, but I didn't check closely.

File: indra/newview/bdfloaterposer.cpp [hunk 9]
```diff
@@ -0,0 +1,2249 @@
...
+    test_row["columns"][COL_ICON]["column"] = "icon";
+    test_row["columns"][COL_ICON]["type"] = "icon";
+    test_row["columns"][COL_ICON]["value"] = getString("icon_category");
+    mJointScrolls[COLLISION_VOLUMES]->addElement(test_row);
+    mJointScrolls[COLLISION_VOLUMES]->deleteSingleItem(0);
+    mJointScrolls[ATTACHMENT_BONES]->addElement(test_row);
+    mJointScrolls[ATTACHMENT_BONES]->deleteSingleItem(0);
+
+    //BD - Collision Volumes
+    for (auto name : cv_names)
```
Comment (Ansariel) [inline indra/newview/bdfloaterposer.cpp:621]: As figured, also not using a const-ref in this file as well...

File: indra/newview/bdfloaterposer.cpp [hunk 10]
```diff
@@ -0,0 +1,2249 @@
...
+//BD - Copy and mirror one side's joints to the other (symmetrize the pose).
+void BDFloaterPoser::onPoseSymmetrize(const LLSD& param)
+{
+    LLVOAvatar* avatar = gDragonAnimator.mTargetAvatar;
+    if (!avatar || avatar->isDead()) return;
+
+    if (!(avatar->getRegion() == gAgent.getRegion())) return;
+
+    LLJoint* joint = nullptr;
+    bool flipped[134] = { false };
```
Comment (Ansariel) [inline indra/newview/bdfloaterposer.cpp:1431]: Oh hello 134 again...

File: indra/newview/llvoavatar.cpp [hunk 11]
```diff
@@ -171,6 +176,7 @@ const LLUUID ANIM_AGENT_PELVIS_FIX = LLUUID("0c5dd2a2-514d-8893-d44d-05beffad208
 const LLUUID ANIM_AGENT_TARGET = LLUUID("0e4896cb-fba4-926c-f355-8720189d5b55");  //"target"
 const LLUUID ANIM_AGENT_WALK_ADJUST = LLUUID("829bc85b-02fc-ec41-be2e-74cc6dd7215d");  //"walk_adjust"
 const LLUUID ANIM_AGENT_PHYSICS_MOTION = LLUUID("7360e029-3cb8-ebc4-863e-212df440d987");  //"physics_motion"
+const LLUUID ANIM_BD_POSING_MOTION     = LLUUID("fd29b117-9429-09c4-10cb-933d0b2ab653");  // <FS> [FIRE-30873]: Poser: "custom_motion"
```
Comment (Ansariel) [inline indra/newview/llvoavatar.cpp:179]: Pretty sure this doesn't exist on OpenSim

**Comment by Ansariel** [comment]
Sorry to say it, but this code is a mess (not your fault though) and I am all against adding this to the viewer. I even received a warning from Rye from Alchemy that it is dangerous and unstable.

**Comment by AngeldarkRaymaker** [comment]
Thank you both for taking the time to have a look.

Beq: sorry about the 'big squares'. That's a texture missing or I've not mapped properly; I see from your screenshot there's more than one missing.

Ansariel: when you say you're 'all against adding this to the viewer', do you mean the code in it's current form, or 'this feature at all'? Not implying anything: I just want to get that clear before continuing.

The risk 10k lines presents can't really be overstated and I'm not going to vouch-safe much there beyond: 'yeah, I use it everyday'. But before I start teasing the code apart (because there are obvious cleavages in what is there, such that this could be added more piecemeal), I'd like to gauge what approach is a good one to work towards.

I'm content to start over: Make PR as small as I can, cutting as much fat as possible from the code, tidy it up, and present a new PR which is a 'minimal' version of Poser (not the animator or the exporter). Thoughts?

What's the team's normal approach to this amount of risk? (just so I can aligned on that)

I don't want to waste time anyone's time, or damage Firestorms reputation.

Either way, it's been a pleasure: thanks again for taking the time with me!

**Comment by beqjanus** [comment]
Firstly, personally I think ~the~ a poser is a great feature. It is something that is requested often, but I have been reluctant to merge this implementation in part for the reasons that Ansa (and Rye) have noted, but more importantly because I think that the implementation is very wrong. The correct way to animate a complex skeleton is through a proxy rig, not through direct manipluation of the bones. The fact that we have collision bones, skeleton bones and attachment bones none of which should be exposed to the end-user (as they are laregly meaningless) immediately says that this is not a user-friendly solution.

Having taken a look at the code and played with it, there are a number of concerns that I have:
* the number of bones exposed
* the use of sliders as a control is sub-optimal
* the total lack visual feedback for selection, there is no display of the bones, there is no way to know (without prior knowledge) what "mChest" does, or perhaps more obviously why are there 4 spines and why do that do weird things

Overall, the inworld solutions such as AnyPose are a superior user experience, though the viewer one offers more precision. My expectation would be a simple user interface for most users, with (perhaps) an "expert" mode that exposes all the bones.

This is just a few notes, based on a quick review. 

I am happy to work with you to break the poser down into smaller chunks and to then work on those. I agree with your perspective, I think the live poser is probably step 1. What I might do is take this into a feature branch so that it can be worked on over time. We have a lot of competing demands at the moment. I think that this needs considerable attention to be a good feature, so it makes sense to branch it and build on it rather than expect it to appear fully formed as a single PR.

**Comment by beqjanus** [comment]
I also had concerns that any work based on the current skeleton would be invalidated by the (near) future GLTF import projects. But following a brief conversation with Runitai Linden yesterday , he confirmed that the existing skelly is lilkely to remain for the forseeable future.

**Comment by AngeldarkRaymaker** [comment]
Thanks Beq! That's a lot for me to think about.
I'm happy to close/decline the PR. I'll fix the missing graphics and keep this synced as I use it at home.

I'll take some time to think about what you've said, there's plenty for me to unpack code-wise: understanding what you mean in terms of classes/methods. I like the idea of re-doing Poser, but I've got a bunch to learn.

I might even try putting a new UI on what's there first? I'll have a think.
Thanks so much!

Random thoughts, feel free to ignore:
Poser is a pig of a UI. It's tedious to learn and days to figure out what does what, and I think about half the controls are un-needed or just repetitious info. I hate zipping up and down the list to find the thing I want to change. Volumes/Bones? They do nothing I care about when I'm making a photo, I don't know why they're there.
I haven't been very successful getting friends to stick with Poser: even though it's so, so good when you learn it...

But I do like the list of bones, for (de)activating lots of bones. It's so nice to change expression, move eyes (the poser can't move eyes for beans, I just use a hud: someone else solved that problem). Re-capture, and then raise the eyebrows a little more, lovely! Save it.

One thing I did pop in there was ctrl-wheel, and shift-ctrl-wheel for sliders, which the Poser floater is awful without. Gross changes, followed by finer changes are important. I also like the sliders because they change function between rotation/position and scale. 'Undo' is also important.
Using azimuth/elevation (such as in the Personal Lighting floater for Sun and Moon) is a nice way to get 3 axis control into two, but I'd want to put ctrl/shift/wheel into that too: one control to change axis (up/down to left/right) while wheeling and another for fine movement. And you can still click-drag it for gross control. Which leads to the final thing about wheel: it's 'relative' control, the change injected is relative to what it is now. A click is telling the control 'go here', and that can be frustrating if your bone zips off to some derpy angle.
NiranV did put something in to not move a slider if one clicks near-enough to the thumb... It shifts the pointer to the middle of the thumb... it's an interesting nice idea.

I love ctrl/alt/shift/wheel, because they're what you're using to shift the camera around the scene while you're posing and shooting.
Anyway, thanks again

**Comment by chanayane** [comment]
I have added the poser from Black dragon about 3 months ago in my fork of Firestorm and I had to fix things with the Firestorm grid component for this to display correctly. Also there are some missing pictures. Feel free to check how I did in my fork.

I did not make a PR about it for the same reasons evoked here, though. It also does not look and feel like a part of Firestorm.

**Comment by beqjanus** [comment]
Hi @chanayane, @AngeldarkRaymaker has been working on this and has been focused on rewriting it to make it less awful. She's made good progress aesthetically and rewritten large parts, I believe. We're going to iterate over the code to make that cleaner. Hopefully we'll end up with something that's stable and has a UI that does not make me twitch.  Another set of eyes on it would be very welcome I am sure. So watch the repo and I'll probably push a feature_branch at some point soon(ish)

**Pull request closed without merging**